File: app/agents/research.py
"""Research agent for market and keyword research."""

from typing import Dict, Any
from langchain_openai import ChatOpenAI
from langchain_core.prompts import ChatPromptTemplate
from app.models.schemas import ResearchReport, KeywordAnalysis, CompetitorInsight
from app.config import DEFAULT_MODEL, TEMPERATURE, OPENAI_API_KEY, PROMPTS_DIR
from app.utils.logging import log_step
import logging
import json


def load_prompt(filename: str) -> str:
    """Load prompt template from file."""
    prompt_path = PROMPTS_DIR / filename
    with open(prompt_path, 'r', encoding='utf-8') as f:
        return f.read()


def research_agent_node(state: Dict[str, Any]) -> Dict[str, Any]:
    """
    Conduct research for the marketing campaign.
    """
    logger = logging.getLogger(f"marketing_system.{state['run_id']}")
    log_step(logger, "Research Agent", "started")
    
    try:
        brief = state['brief']
        
        # Load system prompt
        system_prompt = load_prompt('research.txt')
        
        # Create the LLM with structured output
        llm = ChatOpenAI(
            model=DEFAULT_MODEL,
            temperature=TEMPERATURE,
            api_key=OPENAI_API_KEY
        )
        
        # Create structured output LLM
        structured_llm = llm.with_structured_output(ResearchReport)
        
        # Create prompt
        prompt = ChatPromptTemplate.from_messages([
            ("system", system_prompt),
            ("human", """Conduct comprehensive research for this marketing brief:

Topic: {topic}
Campaign Goal: {campaign_goal}
Target Audience: {target_audience}
Brand Voice: {brand_voice}
{keywords_section}
{context_section}

Provide a detailed research report with keyword analysis, competitor insights, content opportunities, and recommended angles.""")
        ])
        
        # Prepare input
        keywords_section = ""
        if brief.keywords:
            keywords_section = f"Target Keywords: {', '.join(brief.keywords)}"
        
        context_section = ""
        if brief.additional_context:
            context_section = f"Additional Context: {brief.additional_context}"
        
        # Create chain
        chain = prompt | structured_llm
        
        # Invoke
        logger.info("Calling OpenAI for research analysis...")
        research_report = chain.invoke({
            "topic": brief.topic,
            "campaign_goal": brief.campaign_goal,
            "target_audience": brief.target_audience,
            "brand_voice": brief.brand_voice,
            "keywords_section": keywords_section,
            "context_section": context_section
        })
        
        log_step(logger, "Research Agent", "completed")
        logger.info(f"Found {len(research_report.keyword_analysis.primary_keywords)} primary keywords")
        logger.info(f"Analyzed {len(research_report.competitor_insights)} competitors")
        
        return {
            **state,
            "research_report": research_report,
            "status": "research_completed"
        }
        
    except Exception as e:
        log_step(logger, "Research Agent", f"error: {str(e)}")
        
        # Check retry count
        retry_count = state.get('retry_count', 0)
        if retry_count < 1:
            logger.info("Retrying research agent...")
            return {
                **state,
                "retry_count": retry_count + 1,
                "status": "research_retry"
            }
        
        return {
            **state,
            "status": "failed",
            "errors": state.get("errors", []) + [f"Research error: {str(e)}"]
        }

File: src/lib/agents/research.ts
import fs from 'fs/promises';
import path from 'path';
import { openaiChat } from '../llm';
import { ResearchReportSchema } from '../schemas';
import { log, saveArtifact } from '../logging';

function loadPrompt(name: string) {
  const p = path.join(process.cwd(), 'prompts', name);
  return fs.readFile(p, 'utf-8');
}

export async function runResearch(runId: string, brief: any) {
  const system = await loadPrompt('research.txt');
  const user = `Conduct comprehensive research for this marketing brief:\nTopic: ${brief.topic}\nCampaign Goal: ${brief.campaign_goal}\nTarget Audience: ${brief.target_audience}\nBrand Voice: ${brief.brand_voice || ''}\nKeywords: ${brief.keywords?.join(', ') || ''}\nAdditional Context: ${brief.additional_context || ''}\nReturn JSON matching the ResearchReport schema.`;
  await log(runId, 'info', 'Starting research agent');
  const content = await openaiChat([{ role: 'system', content: system }, { role: 'user', content: user }]);
  try {
    const parsed = (await import('../json')).extractJSON(content || '');

    // Normalize model output to match expected schema types
    function normalizeResearch(obj:any) {
      if (!obj || typeof obj !== 'object') return obj;
      const out = JSON.parse(JSON.stringify(obj));

      // keyword_analysis normalization
      if (out.keyword_analysis && typeof out.keyword_analysis === 'object') {
        const ka = out.keyword_analysis;
        if (ka.search_intent && typeof ka.search_intent !== 'string') {
          if (typeof ka.search_intent === 'object') {
            ka.search_intent = ka.search_intent.text || ka.search_intent.intent || JSON.stringify(ka.search_intent);
          } else {
            ka.search_intent = String(ka.search_intent || '');
          }
        }
        if (Array.isArray(ka.primary_keywords)) {
          ka.primary_keywords = ka.primary_keywords.map((k:any)=> typeof k === 'string' ? k : (k && (k.text || k.keyword) ? (k.text || k.keyword) : JSON.stringify(k)));
        }
        if (Array.isArray(ka.secondary_keywords)) {
          ka.secondary_keywords = ka.secondary_keywords.map((k:any)=> typeof k === 'string' ? k : (k && (k.text || k.keyword) ? (k.text || k.keyword) : JSON.stringify(k)));
        }
      }

      // recommended_angles normalization: ensure array of strings
      if (Array.isArray(out.recommended_angles)) {
        out.recommended_angles = out.recommended_angles.map((a:any)=> {
          if (typeof a === 'string') return a;
          if (a && typeof a === 'object') return a.text || a.angle || JSON.stringify(a);
          return String(a);
        });
      }

      return out;
    }

    const normalized = normalizeResearch(parsed);
    const validated = ResearchReportSchema.parse(normalized);
    await saveArtifact(runId, 'research.json', JSON.stringify(validated, null, 2));
    await log(runId, 'info', 'Research completed');
    return validated;
  } catch (e) {
    await log(runId, 'error', `Research parsing/validation error: ${String(e)}`);
    throw e;
  }
}
